upup/pkg/fi/cloud.go

```diff
@@ -63,371 +65,43 @@ type SubnetInfo struct {
-	"centraluseuap":       kops.CloudProviderAzure,
-	"centralusstage":      kops.CloudProviderAzure,
-	"eastasia":            kops.CloudProviderAzure,
-	"eastasiastage":       kops.CloudProviderAzure,
-	"eastus":              kops.CloudProviderAzure,
-	"eastus2":             kops.CloudProviderAzure,
-	"eastus2euap":         kops.CloudProviderAzure,
-	"eastus2stage":        kops.CloudProviderAzure,
-	"eastusstage":         kops.CloudProviderAzure,
-	"europe":              kops.CloudProviderAzure,
-	"francecentral":       kops.CloudProviderAzure,
-	"francesouth":         kops.CloudProviderAzure,
-	"germanynorth":        kops.CloudProviderAzure,
-	"germanywestcentral":  kops.CloudProviderAzure,
-	"global":              kops.CloudProviderAzure,
-	"india":               kops.CloudProviderAzure,
-	"japan":               kops.CloudProviderAzure,
-	"japaneast":           kops.CloudProviderAzure,
-	"japanwest":           kops.CloudProviderAzure,
-	"koreacentral":        kops.CloudProviderAzure,
-	"koreasouth":          kops.CloudProviderAzure,
-	"northcentralus":      kops.CloudProviderAzure,
-	"northcentralusstage": kops.CloudProviderAzure,
-	"northeurope":         kops.CloudProviderAzure,
-	"norwayeast":          kops.CloudProviderAzure,
-	"norwaywest":          kops.CloudProviderAzure,
-	"southafricanorth":    kops.CloudProviderAzure,
-	"southafricawest":     kops.CloudProviderAzure,
-	"southcentralus":      kops.CloudProviderAzure,
-	"southcentralusstage": kops.CloudProviderAzure,
-	"southeastasia":       kops.CloudProviderAzure,
-	"southeastasiastage":  kops.CloudProviderAzure,
-	"southindia":          kops.CloudProviderAzure,
-	"switzerlandnorth":    kops.CloudProviderAzure,
-	"switzerlandwest":     kops.CloudProviderAzure,
-	"uaecentral":          kops.CloudProviderAzure,
-	"uaenorth":            kops.CloudProviderAzure,
-	"uk":                  kops.CloudProviderAzure,
-	"uksouth":             kops.CloudProviderAzure,
-	"ukwest":              kops.CloudProviderAzure,
-	"unitedstates":        kops.CloudProviderAzure,
-	"westcentralus":       kops.CloudProviderAzure,
-	"westeurope":          kops.CloudProviderAzure,
-	"westindia":           kops.CloudProviderAzure,
-	"westus":              kops.CloudProviderAzure,
-	"westus2":             kops.CloudProviderAzure,
-	"westus2stage":        kops.CloudProviderAzure,
-	"westusstage":         kops.CloudProviderAzure,
-}
-
 // GuessCloudForZone tries to infer the cloudprovider from the zone name
 // Ali has the same zoneNames as AWS in the regions outside China, so if use AliCloud to install k8s in the regions outside China,
 // the users need to provide parameter "--cloud". But the regions inside China can be easily identified.
 func GuessCloudForZone(zone string) (kops.CloudProviderID, bool) {
-	c, found := zonesToCloud[zone]
-	return c, found
+
+	structuredProviders := map[kops.CloudProviderID]string{
+		kops.CloudProviderAWS: "^[a-z]{2}-[a-z\\-]+?-[0-9]{1,2}[a-z]$",
```

Comment:
ap-south-1a and ab-south-1a will both match in this case. I am assuming there will be further checks validating the zone and region specified?

Reply:
My understanding is that this method tries to guess the cloud provider when --cloud has not been provided, but --zones has been provided.  Later, when it actually tries to communicated with aws it will error that ab-south-1a does not exist.  

I think the _main_ drawback of this approach is the regex for azure being so generic.  It will assume azure instead of asking the user to specify a cloud provider in the case that none of the other providers match and the zone is alphanumeric.  So it could guess Azure too often.  But again, if that is wrong, the user can specify --cloud with the correct provider.  

This may be a good temporary solution to deprecating allowing --cloud to be optional.  This goes away altogether if --cloud becomes required.